### PR TITLE
Porting to cygwin

### PR DESCRIPTION
**Pull request by tinysun212**
I made several changes set for cygwin64 port.

The standard library can be used and It is possible to compile and run or interpret a simple hello swift source, but REPL, swift-build and swift-lldb does not work.

Some of these would bad result in other platforms (ex. OS X, Linux) because I didn't check, but I think it would easily fixed.

### CODE REVIEW

File: include/swift/Basic/Dwarf.h [hunk 1]
```diff
@@ -21,7 +21,7 @@
 
 namespace swift {
   /// The DWARF version emitted by the Swift compiler.
-  const unsigned DWARFVersion = 3;
+  const unsigned DWARFVersion = 4;
```
Comment (gribozavr) [inline include/swift/Basic/Dwarf.h]: Why was this required?  /cc @adrian-prantl

File: lib/Basic/LangOptions.cpp [hunk 2]
```diff
@@ -101,6 +101,8 @@ std::pair<bool, bool> LangOptions::setTarget(llvm::Triple triple) {
     addTargetConfigOption("os", "Linux");
   else if (triple.isOSFreeBSD())
     addTargetConfigOption("os", "FreeBSD");
+  else if (triple.isOSWindows())
+    addTargetConfigOption("os", "Windows");
```
Comment (gribozavr) [inline lib/Basic/LangOptions.cpp]: For this, please add tests like `test/BuildConfigurations/x64LinuxTarget.swift` for every new supported triple.

File: lib/Driver/ToolChains.cpp [hunk 3]
```diff
@@ -1112,6 +1112,7 @@ toolchains::GenericUnix::constructInvocation(const LinkJobAction &job,
         Twine("@") + OutputInfo.getPrimaryOutputFilename()));
   }
 
+#if !defined(__CYGWIN__)
```
Comment (gribozavr) [inline lib/Driver/ToolChains.cpp]: We don't use `#if` in the compiler.  Please create a separate toolchain class for Cygwin, if it requires customization, or create customization parameters in the `GenericUnix` constructor.

File: lib/IRGen/GenDecl.cpp [hunk 4]
```diff
@@ -1881,6 +1881,9 @@ llvm::Constant *IRGenModule::emitProtocolConformances() {
   case llvm::Triple::ELF:
     sectionName = ".swift2_protocol_conformances";
     break;
+  case llvm::Triple::COFF:  // for cygwin
```
Comment (gribozavr) [inline lib/IRGen/GenDecl.cpp]: Please drop the comment.  This code will automatically get activated in more cases than just Cygwin, so it is misleading.

File: lib/IRGen/IRGenModule.cpp [hunk 5]
```diff
@@ -693,6 +693,7 @@ void IRGenModule::emitAutolinkInfo() {
     }
     break;
   }
+  case llvm::Triple::COFF:  // for cygwin
```
Comment (gribozavr) [inline lib/IRGen/IRGenModule.cpp]: Please drop the comment, too.

Comment (gribozavr) [inline lib/IRGen/IRGenModule.cpp]: Is the section name `.swift1_autolink_entries` too long for COFF?  How does autolinking work?

File: stdlib/public/CMakeLists.txt [hunk 6]
```diff
@@ -12,6 +12,8 @@ if(CXX_SUPPORTS_EXIT_TIME_DESTRUCTORS_WARNING)
   list(APPEND SWIFT_CORE_CXX_FLAGS "-Wexit-time-destructors")
 endif()
 
+list(APPEND SWIFT_CORE_CXX_FLAGS "-mcmodel=large")
```
Comment (gribozavr) [inline stdlib/public/CMakeLists.txt]: Please scope this option to Cygwin.

File: stdlib/public/core/CMakeLists.txt [hunk 7]
```diff
@@ -165,6 +165,13 @@ if(CMAKE_SYSTEM_NAME STREQUAL "Linux")
       ${BSD_LIBRARIES})
 endif()
 
+if(CMAKE_SYSTEM_NAME STREQUAL "CYGWIN")
+  execute_process( COMMAND "cygpath" "-u" "$ENV{SYSTEMROOT}"
```
Comment (gribozavr) [inline stdlib/public/core/CMakeLists.txt]: Please drop the space before `COMMAND`

File: stdlib/public/core/Print.swift [hunk 8]
```diff
@@ -28,6 +28,8 @@ public func print(
   separator: String = " ",
   terminator: String = "\n"
 ) {
+  // Adhoc fix for 'crash at hook(output.left)' in cygwin
+  let _playgroundPrintHook : ((String)->Void)? = {_ in () }
```
Comment (gribozavr) [inline stdlib/public/core/Print.swift]: This would break playgrounds on OS X.  Please use `#if` to scope the workaround to Cygwin, file an issue in the bugtracker to implement a proper fix, and reference it here.

File: stdlib/public/runtime/CygwinPort.cpp [hunk 9]
```diff
@@ -0,0 +1,131 @@
+#include <stdint.h>
```
Comment (gribozavr) [inline stdlib/public/runtime/CygwinPort.cpp]: Please add a license header, and run clang-format over the file.

File: stdlib/public/runtime/CygwinPort.cpp [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+			*section_size = VirtualSize;
+			return  (uint8_t *)handle + VirtualAddress;
+		}
+		section_header += kSectionRecordSize;
+	}
+
+	return  nullptr;
+}
+
+namespace std {
```
Comment (gribozavr) [inline stdlib/public/runtime/CygwinPort.cpp]: Defining entities in `namespace std` is undefined behavior, please don't do it.

File: stdlib/public/runtime/Once.cpp [hunk 11]
```diff
@@ -31,8 +31,10 @@ static_assert(std::is_same<swift_once_t, dispatch_once_t>::value,
 // The compiler generates the swift_once_t values as word-sized zero-initialized
 // variables, so we want to make sure swift_once_t isn't larger than the
 // platform word or the function below might overwrite something it shouldn't.
+#if !defined(__CYGWIN__)
```
Comment (gribozavr) [inline stdlib/public/runtime/Once.cpp]: Is this failing on Cygwin?  Why?  This is an important check.

File: stdlib/public/runtime/CygwinPort.cpp [hunk 12]
```diff
@@ -0,0 +1,131 @@
...
+#include <psapi.h>
+#include <mutex>
+
+struct dl_phdr_info
+{
+	void			 *dlpi_addr;
+	const char       *dlpi_name;
+};
+
+int dl_iterate_phdr(int(*callback)(struct dl_phdr_info *info, size_t size, void *data),
```
Comment (gribozavr) [inline stdlib/public/runtime/CygwinPort.cpp]: Please don't export symbols with names other than `swift_*` from the runtime.  I understand that you need a shim layer to make existing code work with WinAPI.  Let's create one without creating extra exports.

File: stdlib/public/stubs/Stubs.cpp [hunk 13]
```diff
@@ -29,12 +35,14 @@
 #include <cstdio>
 #include <cstdlib>
 #include <cstring>
-#include <xlocale.h>
+//#include <xlocale.h>
```
Comment (gribozavr) [inline stdlib/public/stubs/Stubs.cpp]: This will break OS X.

File: stdlib/public/stubs/Stubs.cpp [hunk 14]
```diff
@@ -20,6 +20,12 @@
 #define _WITH_GETLINE
 #endif
 
+#if defined(__CYGWIN__)
+#include <stdio.h>
+typedef  void *locale_t;
+#define uselocale(x)  nullptr
```
Comment (gribozavr) [inline stdlib/public/stubs/Stubs.cpp]: Setting the locale to C is an important part of the Swift's floating point parsing semantics, to ensure that numbers are parsed the same way regardless of the locale.  Please don't drop it.

File: stdlib/public/stubs/UnicodeNormalization.cpp [hunk 15]
```diff
@@ -129,8 +129,8 @@ int32_t _swift_stdlib_unicode_compare_utf16_utf16(const uint16_t *LeftString,
                                                   const uint16_t *RightString,
                                                   int32_t RightLength) {
   return ucol_strcoll(GetRootCollator(),
-    LeftString, LeftLength,
-    RightString, RightLength);
+    (const UChar *)LeftString, LeftLength,
+    (const UChar *)RightString, RightLength);
```
Comment (gribozavr) [inline stdlib/public/stubs/UnicodeNormalization.cpp]: Why were changes required here?  What is `UChar` on Windows?

File: utils/build-script-impl [hunk 16]
```diff
@@ -763,6 +766,12 @@ case "$(uname -s -m)" in
             "freebsd-x86_64"
         )
         ;;
+		
+	CYGWIN_NT-10.0\ x86_64)
```
Comment (gribozavr) [inline utils/build-script-impl]: Indentation is off on this line.

File: utils/build-script-impl [hunk 17]
```diff
@@ -849,6 +858,12 @@ case "$(uname -s -m)" in
         )
         ;;
 
+	CYGWIN_NT-10.0\ x86_64)
```
Comment (gribozavr) [inline utils/build-script-impl]: Indentation is off on this line.

**Comment by gribozavr** [comment]
@tinysun212 Thank you for your work!  I have left a few comments, unfortunately some of issues are showstoppers.  I would be happy to discuss possible solutions.

File: lib/SIL/Projection.cpp [hunk 18]
```diff
@@ -1976,12 +1976,12 @@ replaceValueUsesWithLeafUses(SILBuilder &Builder, SILLocation Loc,
...
-    for (auto *_N : Worklist) {
-      DEBUG(llvm::dbgs() << "        Type: " << _N->getType()
+    for (auto *_work : Worklist) {
+      DEBUG(llvm::dbgs() << "        Type: " << _work->getType()
                    << "; Complete: "
-                   << (AggBuilderMap.isComplete(_N)? "yes" : "no")
+                   << (AggBuilderMap.isComplete(_work)? "yes" : "no")
                    << "; Invalidated: "
-            << (AggBuilderMap.isInvalidated(_N)? "yes" : "no") << "\n");
+            << (AggBuilderMap.isInvalidated(_work)? "yes" : "no") << "\n");
```
Comment (gribozavr) [inline lib/SIL/Projection.cpp]: You can submit this part as a separate pull request if you like, I'll merge it immediately.  Not using reserved names is obvious goodness.

File: stdlib/public/core/Print.swift (hunk 8 above)
Comment (gribozavr) [inline stdlib/public/core/Print.swift]: Also, the Cygwin workaround would be:

```
let _playgroundPrintHook: ((String)->Void)? = nil
```

... it is cheaper.

File: stdlib/public/runtime/CygwinPort.cpp [hunk 19]
```diff
@@ -0,0 +1,131 @@
...
+					void *data)
+{
+	DWORD proc_id = GetCurrentProcessId();
+	HANDLE proc_handle = OpenProcess(PROCESS_QUERY_INFORMATION | PROCESS_VM_READ,
+									 FALSE, proc_id);
+	if (NULL == proc_handle)
+		return 0;
```
Comment (gribozavr) [inline stdlib/public/runtime/CygwinPort.cpp]: Please crash the program with `swift::fatalError` if any required system call fails.  The program definitely won't run correctly if we can't read conformance tables, so there is no point in continuing.

File: stdlib/public/runtime/CygwinPort.cpp [hunk 20]
```diff
@@ -0,0 +1,131 @@
...
+
+	int last_ret = 0;
```
Comment (gribozavr) [inline stdlib/public/runtime/CygwinPort.cpp]: Please follow the LLVM coding style in variable names.

File: stdlib/public/runtime/CygwinPort.cpp [hunk 21]
```diff
@@ -0,0 +1,131 @@
...
+
+	HMODULE modules[1024];
+	DWORD used_size;
+
+	if (EnumProcessModules(proc_handle, modules, sizeof(modules), &used_size))
```
Comment (gribozavr) [inline stdlib/public/runtime/CygwinPort.cpp]: Please crash the program with `swift::fatalError` if this fails.  Please also implement the retry loop as described in MSDN docs to find the proper size of the `modules` array.

File: stdlib/public/runtime/CygwinPort.cpp [hunk 22]
```diff
@@ -0,0 +1,131 @@
...
+	{
+		for (unsigned int i = 0; i<used_size/sizeof(HMODULE); i++)
+		{
+			char mod_name[MAX_PATH];
+
+			if (GetModuleFileNameExA(proc_handle, modules[i], mod_name,
```
Comment (gribozavr) [inline stdlib/public/runtime/CygwinPort.cpp]: What are the conditions under which this call can fail?  If those are obscure system failure cases, let's crash with `swift::fatalError` instead.

File: include/swift/Basic/Dwarf.h (hunk 1 above)
Comment (tinysun212) [inline include/swift/Basic/Dwarf.h]: My dev environment was cygwin 64bit 2.3.1, and clang version was 3.5.2.
The dwarf version 4 is used as default and the linker displayed errors when
compiled with dwarf version 3.

-----
$ clang++ hello.cpp -gdwarf-3
/tmp/hello-c8fc34.o:(.debug_info+0x16): relocation truncated to fit:
R_X86_64_32 against `.debug_line'
collect2: error: ld returned 1 exit status
clang: error: linker (via gcc) command failed with exit code 1 (use -v to
see invocation)
-----

So swiftc -c -g must generate in dwarf ver 4.

Is this should be changed to like this?

-----
#if __CYGWIN__
  const unsigned DWARFVersion = 4;
#else
  const unsigned DWARFVersion = 3;
#fi
-----


2016-01-19 9:10 GMT+09:00 Dmitri Gribenko <notifications@github.com>:

Comment (gribozavr) [inline include/swift/Basic/Dwarf.h]: I see.  Let's see what @adrian-prantl says, possibly this change would need to be scoped to Cygwin.

File: stdlib/public/runtime/Once.cpp (hunk 11 above)
Comment (tinysun212) [inline stdlib/public/runtime/Once.cpp]: In cygwin 64bit v2.3.1, clang v3.5.2, the sizeof(std::once_flag) is 16 and sizeof(void*) is 8.

Comment (gribozavr) [inline stdlib/public/runtime/Once.cpp]: Well, then you need to keep the assertion and adjust `swift::swift_once` -- otherwise the initialization won't be correct.

File: stdlib/public/runtime/ProtocolConformance.cpp [hunk 23]
```diff
@@ -340,7 +380,7 @@ static void _initializeCallbacksToInspectDylib() {
   // Dyld will invoke this on our behalf for all images that have already
   // been loaded.
   _dyld_register_func_for_add_image(_addImageProtocolConformances);
-#elif defined(__ELF__)
+#elif defined(__ELF__) || defined(__CYGWIN__)
```
Comment (lhoward) [inline stdlib/public/runtime/ProtocolConformance.cpp]: If this would work on Win32 as well then checking for WIN32 might be better.

File: stdlib/public/runtime/ProtocolConformance.cpp [hunk 24]
```diff
@@ -332,6 +347,31 @@ static int _addImageProtocolConformances(struct dl_phdr_info *info,
...
+    getSectionDataPE(handle, SWIFT_PROTOCOL_CONFORMANCES_SECTION,
+                     &conformancesSize);
+
+  if (!conformances) {
+    // if there are no conformances, don't hold this handle open.
+    dlclose(handle);
+    return 0;
+  }
+
+  _addImageProtocolConformancesBlock(conformances, conformancesSize);
```
Comment (lhoward) [inline stdlib/public/runtime/ProtocolConformance.cpp]: Also current master needs similar code in MetadataLookup.cpp.

File: lib/IRGen/IRGenModule.cpp (hunk 5 above)
Comment (tinysun212) [inline lib/IRGen/IRGenModule.cpp]: The cygwin tool chain covers the long section names and generates long section name contained binaries. The COFF format has a separated string table for names longer than 8, but the table is dropped when loaded to system memory. At run-time, the long name section is exist in memory but can not be identified.

Without 'swift2_protocol_conformances', print("Hello") didn't work, so I had to modify it.
But I don't know that how 'swift1_autolink_entries' works and didn't touch it.

I think that If it must be accessed at run-time, it also has to have a shorter name.
But I hope keep it unchanged now and resolve it in another commit later.

Comment (gribozavr) [inline lib/IRGen/IRGenModule.cpp]: The autolinking section is used by the `swift-autolink-extract` to enable autolinking of the dependent libraries.

File: stdlib/public/stubs/UnicodeNormalization.cpp (hunk 15 above)
Comment (tinysun212) [inline stdlib/public/stubs/UnicodeNormalization.cpp]: Without casting, compile error occurred.

In cygwin 64bit, I could get the definition with clang++ -E.
typedef wchar_t UChar;
and by simple code, sizeof(UChar) was 2.

I found where it defined.
------------
/usr/include/unicode/umachine.h

/**
 * \var UChar
 * Define UChar to be UCHAR_TYPE, if that is #defined (for example, to char16_t),
 * or wchar_t if that is 16 bits wide; always assumed to be unsigned.
 * If neither is available, then define UChar to be uint16_t.
 *
 * This makes the definition of UChar platform-dependent
 * but allows direct string type compatibility with platforms with
 * 16-bit wchar_t types.
 *
 * @stable ICU 4.4
 */
#if defined(UCHAR_TYPE)
    typedef UCHAR_TYPE UChar;
/* Not #elif U_HAVE_CHAR16_T -- because that is type-incompatible with pre-C++11 callers
    typedef char16_t UChar;  */
#elif U_SIZEOF_WCHAR_T==2
    typedef wchar_t UChar;
#elif defined(__CHAR16_TYPE__)
    typedef __CHAR16_TYPE__ UChar;
#else
    typedef uint16_t UChar;
#endif

File: stdlib/public/runtime/ProtocolConformance.cpp (hunk 23 above)
Comment (tinysun212) [inline stdlib/public/runtime/ProtocolConformance.cpp]: Do you mean like this ?
```
|| defined(__WIN32__)
```

File: stdlib/public/stubs/UnicodeNormalization.cpp (hunk 15 above)
Comment (gribozavr) [inline stdlib/public/stubs/UnicodeNormalization.cpp]: I see.  This is complicated.  Could you check that casting `uint16_t *` to `wchar_t *` is valid from the strict aliasing perspective?  We don't want miscompiles.  Otherwise, I'm OK with the change.

Comment (tinysun212) [inline stdlib/public/stubs/UnicodeNormalization.cpp]: If I understand you correctly, this is helpful.
```
#include <stdint.h>
#include <stdio.h>

int main()
{
        uint16_t u16_var = 1234;
        wchar_t  *wc_var = (wchar_t *)&u16_var;
        printf("%p\n", wc_var);

        return 0;
}

$ clang++ wchar_test.cpp -fstrict-aliasing -Wall
// no error, no warning
```

Comment (gribozavr) [inline stdlib/public/stubs/UnicodeNormalization.cpp]: Thank you for checking, but this does not prove that the cast is correct.  In fact, clang does not warn on this code on OS X, where `sizeof(wchar_t) == 4`, even if I dereference the pointer with `printf("%d\n", *wc_var);`.

My interpretation of N4431 3.10/10 is that dereferencing that pointer would be a violation of strict aliasing.  @rjmccall?

File: include/swift/Basic/Dwarf.h (hunk 1 above)
Comment (adrian-prantl) [inline include/swift/Basic/Dwarf.h]: Upgrading the emitted DWARF version has to be done very carefully, because there are many different DWARF consumers besides the debugger that all need to be tested for compatibility and upgraded. There is work underway to move the Darwin default for both clang and swiftc to DWARF 4, but the qualification process is not yet complete.

The best solution is to implement -gdwarf-3 / -gdwarf-4 / -gdwarf-5 driver options that translate to -dwarf-version=<N> frontend options analogous to the clang driver and make the default depend on the target platform.

File: lib/Basic/LangOptions.cpp (hunk 2 above)
Comment (tinysun212) [inline lib/Basic/LangOptions.cpp]: I added x64WindowsTarget.swift in later commit.

**Comment by gribozavr** [comment]
@tinysun212 Are you waiting for feedback or are you still working on the patch?

**Comment by gribozavr** [comment]
@tinysun212 Oh, and another thing is that we don't commit to release branches directly.  We first commit to master, and then made a decision to cherry-pick to release if the change is considered safe enough.

Could you change the PR to merge to `master` instead?

**Comment by tinysun212** [comment]
I was working on the patch.

**Comment by gribozavr** [comment]
OK.  Let me know when it is ready for review!

**Comment by tinysun212** [comment]
I read this message now, after a commit an hour ago. 
I have some questions. Could you tell me answers?

1) Do you mean I rebase to master branch at my repo and then PR?

if YES,
2) commits should be retained as now or merged or separated to several PRs ?
3) PR branch should be start at top or near top of the master branch?

**Comment by gribozavr** [comment]
This repository generally prefers to rebase to the top of the master branch.  You can split the PR into separate ones if you feel like the change is getting too large.  I think the current size is good for one PR, but if you're going to add a lot more code to this PR, let's split it.

**Comment by tinysun212** [comment]
I recreated same PR on master branch an hour ago. I think this PR must be removed.

**Comment by gribozavr** [comment]
Closing this PR in favor of #1108.
